### ISSUE TITLE
Job shop example: Can Constraints solver solve such a problem?

### ISSUE THREAD

**Issue by maxnilz** [google/or-tools#945]
Hi, I'm very curious about this situation:
What if we replace the constraints with: 
1. `Each job can run on any machine`
2. `Each machine have a different cost rate per seconds`
3. `Get a schedule sequence that finished all jobs with a minimize the cost`

let's assume we have 100 machines and 3000 jobs.  then the solution space will be very huge: 3000!.

How the Constraints solver solve such a problem?

**Comment by lperron** [comment]
Is each job a unique job or multiple sub-tasks?
Do you have precedences?
Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00

**Comment by maxnilz** [comment]
Thanks for your quick reply, and sorry for giving an ambiguous problem definition since lack of thinking.

Hmm...let's say(this might be a very difficult problem inspired by the original `job shop problem`):

1. `A machine can only work on one job at a time`
2. `A job once started, must run to completion`
3. `Each machine have a different cost rate`
4. `Each job can run on any machine with a specific time duration, and different job have different running duration`
5. `A job has a specific profit rate, different jobs have different profit rate`
5. `Each job is a unique job and have no sub-tasks`
6. `Every job have same priority`
7. `Two consecutive jobs on the same machine are dependent on each other. When one job is finished and the next one is ready to start, a period of warm-up time is required. The warm-up time of different jobs on the same machine is different`

And assume we have 100 machines can be used at the same time, 3000 such jobs need to schedule, we need to find a schedule sequence in a time duration(e.g: 30days) that `maximize the total profit`.

Because the problem might be more generic and the solution space will be very huge, I'm curious the internal mechanism of `or-tools` and want to know how we model this problem with this lib.

**Comment by lperron** [comment]
1) standard no_overlap constaint.
2) standard IntervalVar.
3) when creating optional copies of the master interval, assign a cost to
each Boolean presence variable. See examples/cpp/jobshop_sat.cc
4) the duration depends on the job, on the machine, or both.  Anyway, all
are easy, See the same jobshop_sat.cc
5) What is profit? how it is used in the code?
6) easier
7) ok.
8) Setup time depends on the job, or on the job and its predecessor? If it
depends on both, it complicates the problem. If it depends just on the job,
just extend the duration of the job by its setup time.

I still miss the objective part.
Cost depends on which jobs are assigned to which machine?
How does profit work? If all jobs have a constant profit, and we schedule
all of them, then profit is constant.
Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00

**Comment by lperron** [comment]
If the both profit_rate_per_second and job_duration_in_second are fixed, you can simplify and have a fixed gain per job.

How diverse are the setup cost?

**Comment by maxnilz** [comment]
Thanks for your quick reply very much

>If the both profit_rate_per_second and job_duration_in_second are fixed, you can simplify and have a fixed gain per job.

Yes, you are right. the profit of a job can simplify to a fixed number, and different have a different profit number.

> How diverse are the setup cost?
The setup cost of a job on a machine = `job_set_up_time * machine_cost_rate_per_second`
Although each machine has a different `machine_cost_rate_per_second`, the jobs running on the same machine has the same `machine_cost_rate_per_second`, so the setup-cost diversity mainly depends on the diverse `setup-time` of the jobs
And about the job `setup-time`, Let's say: 
1. each machine has a `setup-speed` and different machines have different `setup-speed`, and two jobs running on the same machine: Job1 and Job2 (Job1 is the predecessor node of Job2) have a `setup-distance`, then the `setup-time = setup-distance/setup-speed of Job2`
2. `Setup-distance` between jobs can be obtained by querying a `predefined table`

**Comment by lperron** [comment]
Thanks

I meant what is the range of setup_times compared to job_durations??
Can we just add max(setup_times) to each job as a first approximation?
Is there a zero setup time between jobs of the same 'family'?

The underlying idea is to compute the assignment of jobs to machines first,
them compute the optimal sequence on each machine.
Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00

**Comment by maxnilz** [comment]
> I meant what is the range of setup_times compared to job_durations??

Job_duration and setup-time are independent of each other, job_duration is an attribute of a job and the job setup-time depends on the previous job on the same machine, both of them are a positive number.

> Can we just add max(setup_times) to each job as a first approximation?

Maybe can't,  the setup-time might be the key feature, as you mentioned before,  and it complicates the problem when the setup time depends on job, job's predecessor and machine that will be scheduled.

> Is there a zero setup time between jobs of the same 'family'?

Yes, maybe it can be take it into account also.

**Comment by maxnilz** [comment]
> The underlying idea is to compute the assignment of jobs to machines first,
then compute the optimal sequence on each machine.

Yes, but the assignments solution space might be very huge.  for example, if we have 3000 jobs then the assignments solution seuqnce space might be 3000!(factorial)

**Comment by lperron** [comment]
No, I meant, if you can approximate setup times, you can create a simple
assignment problem that will compute how many of each job type to assign to
each machine under capacity constraints.
Once this is done, for each machine, you pack jobs of the same type to
minimize setup types, and you push as many jobs as possible on that machine.
Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00

**Comment by maxnilz** [comment]
Thanks very much for your sharing.

I'll try to solve this problem although I'm not sure what's the exact mean of  `
you can create a simple assignment problem that will compute how many of each job type to assign to
each machine under capacity constraints. Once this is done, for each machine, you pack jobs of the same type to minimize setup types, and you push as many jobs as possible on that machine.`

And if success, then I'll share it as another example of `job shop scheduling` 

Since I'm new to `or-tools` and lack of experience in solving NP problem, It will be very helpful If you have more details about how to model it by using `or-tools` 

Start by looking into `examples/cpp/jobshop_sat.cc`

**Comment by lperron** [comment]
https://developers.google.com/optimization/assignment/simple_assignment
Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00

**Comment by tuosor** [comment]
Hello,

I'm working on a similar problem. In this case, there is a setup time that is dependent on both the current job and the previous one in each machine. And, as lperron commented, there is "a zero setup time between jobs of the same 'family'".

How am I able to accomplish this kind of addition? This was not answered in the above thread.
